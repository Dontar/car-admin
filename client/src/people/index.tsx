import {
    Datagrid, DateField, DateInput, Edit, EditProps, List, ListProps, ReferenceArrayField,
    ReferenceArrayInput,
    Show, ShowProps, SimpleForm, SimpleShowLayout, TextField, TextInput
} from "react-admin";

export const PersonList = (props: ListProps) => (
    <List {...props}>
        <Datagrid rowClick="edit">
            {/* <TextField source="id" /> */}
            <TextField source="client_name" />
            <TextField source="egn" />
            {/* <TextField source="representative" /> */}
            <TextField source="phone" />
            <TextField source="mobile" />
            {/* <TextField source="region_name" /> */}
            {/* <TextField source="municipality_name" /> */}
            <TextField source="city" />
            {/* <DateField source="postcode" /> */}
            <TextField source="street_name" />
            <TextField source="street_no" />
            {/* <TextField source="blok" /> */}
            {/* <TextField source="vhod" /> */}
            {/* <TextField source="apartment" /> */}
            {/* <DateField source="floor" /> */}
            <ReferenceArrayField source="car_ids" reference="cars">
                <TextField source="dkn" />
            </ReferenceArrayField>
        </Datagrid>
    </List>
);

export const PersonShow = (props: ShowProps) => (
    <Show {...props}>
        <SimpleShowLayout>
            <TextField source="id" />
            <TextField source="client_name" />
            <TextField source="egn" />
            <TextField source="representative" />
            <DateField source="phone" />
            <TextField source="mobile" />
            <TextField source="region_name" />
            <TextField source="municipality_name" />
            <TextField source="city" />
            <DateField source="postcode" />
            <TextField source="street_name" />
            <DateField source="street_no" />
            <DateField source="blok" />
            <DateField source="vhod" />
            <DateField source="apartment" />
            <DateField source="floor" />
            <ReferenceArrayField source="car_ids" reference="cars">
                <TextField source="id" />
            </ReferenceArrayField>
        </SimpleShowLayout>
    </Show>
);

export const PersonEdit = (props: EditProps) => (
    <Edit {...props}>
        <SimpleForm>
            <TextInput source="id" />
            <TextInput source="client_name" />
            <TextInput source="egn" />
            <TextInput source="representative" />
            <DateInput source="phone" />
            <TextInput source="mobile" />
            <TextInput source="region_name" />
            <TextInput source="municipality_name" />
            <TextInput source="city" />
            <DateInput source="postcode" />
            <TextInput source="street_name" />
            <DateInput source="street_no" />
            <DateInput source="blok" />
            <DateInput source="vhod" />
            <DateInput source="apartment" />
            <DateInput source="floor" />
            <ReferenceArrayInput source="car_ids" reference="cars">
                <TextInput source="id" />
            </ReferenceArrayInput>
        </SimpleForm>
    </Edit>
);