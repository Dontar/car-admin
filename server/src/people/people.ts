import { getDB } from "../connection";
import rawSql, { in as $in, select } from 'sql-bricks-sqlite';
import { GetOneParams, GetOneResult, GetParams, GetResult, IPerson } from "../share/models";
import { isOne, prepareParams, stringIdsToArray } from "../share/utils";


export async function getPersons(params: GetOneParams): Promise<GetOneResult<IPerson>>;
export async function getPersons(params: Partial<GetParams>): Promise<GetResult<IPerson>>;
export async function getPersons(params: Partial<GetParams> | GetOneParams): Promise<GetResult<IPerson> | GetOneResult<IPerson>> {
    const db = await getDB();

    const sql = select(
        'c.ID as id',
        'c.CLIENT_NAME as client_name',
        'c.EGN as egn',
        'c.REPRESENTATIVE as representative',
        'c.PHONE as phone',
        'c.MOBILE as mobile',
        'c.REGION_NAME as region_name',
        'c.MINUCIPALITY_NAME as municipality_name',
        'c.CITY as city',
        'c.POSTCODE as postcode',
        'c.STREET_NAME as street_name',
        'c.STREET_NO as street_no',
        'c.BLOK as blok',
        'c.VHOD as vhod',
        'c.APARTMENT as apartment',
        'c.FLOOR as floor',
        'group_concat(vc.VEHICLE_ID) as car_ids'
    ).from('clients c')
        .leftJoin('vc', { 'vc.CLIENT_ID': 'c.ID' })
        .where('c.CLIENT_TYPE', rawSql(1))
        .groupBy('c.ID');

    let total: number | undefined = 1030889;
    if (isOne(params)) {
        sql.where({ 'c.id': params.id });
    } else {
        if (params.ids) {
            sql.where($in('c.id', ...params.ids));
        } else {
            /* total = */ await prepareParams(params, sql, db);
        }
    }

    const { text, values } = sql.toParams();
    if (isOne(params)) {
        const row = await db.get<IPerson>(text, values);
        return {
            data: stringIdsToArray(row!)
        };
    } else {
        const rows = await db.all<IPerson[]>(text, values);
        return {
            data: stringIdsToArray(rows),
            total
        };
    }
}
