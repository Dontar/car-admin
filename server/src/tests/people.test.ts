import { config } from 'dotenv';
import { getPersons } from "../people/people";
config();

describe('Test persons data retrieval', () => {

    it('should get a person from db', async () => {
        const data = await getPersons({
            id: 9219
        });
        expect(data).toBeDefined();
        expect(data).toMatchSnapshot();
    });

    it('should get 2 persons from db', async () => {
        const data = await getPersons({
            ids: [9219, 9451]
        });
        expect(data.data?.length).toBe(2);
        expect(data).toMatchSnapshot();
    });

    it('should get persons from db with pagination', async () => {
        const data = await getPersons({
            pagination: { page: 1, perPage: 2 }
        });
        expect(data.data?.length).toBe(2);
        expect(data).toMatchSnapshot();
    });

    it('should get persons from db with sorting', async () => {
        const data = await getPersons({
            pagination: { page: 1, perPage: 2 },
            sort: { field: 'id', order: 'ASC' }
        });
        expect(data.data?.length).toBe(2);
        expect(data).toMatchSnapshot();
    });
});
