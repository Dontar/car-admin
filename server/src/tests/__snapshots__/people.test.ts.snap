// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test persons data retrieval should get 2 persons from db 1`] = `
Object {
  "data": Array [
    Object {
      "apartment": "2",
      "blok": "18",
      "car_ids": null,
      "city": "СИЛИСТРА",
      "client_name": "First Middle Last",
      "egn": "6310305565",
      "floor": "1",
      "id": 9219,
      "mobile": "0891043493",
      "municipality_name": "Силистра",
      "phone": "02830123",
      "postcode": "7500",
      "region_name": "СИЛИСТРА",
      "representative": "",
      "street_name": "ГЕНЕРАЛ  СКОБЕЛЕВ",
      "street_no": "",
      "vhod": "А",
    },
    Object {
      "apartment": "7",
      "blok": "9",
      "car_ids": null,
      "city": "СОФИЯ",
      "client_name": "Николинка Стефанова Цолова",
      "egn": "5507181593",
      "floor": "4",
      "id": 9451,
      "mobile": "0897527002",
      "municipality_name": "Столична",
      "phone": "0888998899",
      "postcode": "1000",
      "region_name": "София (столица)",
      "representative": "",
      "street_name": "ЛЮБЛАНА",
      "street_no": "1",
      "vhod": "a",
    },
  ],
  "total": 1030889,
}
`;

exports[`Test persons data retrieval should get a person from db 1`] = `
Object {
  "data": Object {
    "apartment": "2",
    "blok": "18",
    "car_ids": null,
    "city": "СИЛИСТРА",
    "client_name": "First Middle Last",
    "egn": "6310305565",
    "floor": "1",
    "id": 9219,
    "mobile": "0891043493",
    "municipality_name": "Силистра",
    "phone": "02830123",
    "postcode": "7500",
    "region_name": "СИЛИСТРА",
    "representative": "",
    "street_name": "ГЕНЕРАЛ  СКОБЕЛЕВ",
    "street_no": "",
    "vhod": "А",
  },
}
`;

exports[`Test persons data retrieval should get persons from db with pagination 1`] = `
Object {
  "data": Array [
    Object {
      "apartment": "0",
      "blok": "0",
      "car_ids": null,
      "city": "РУСЕ",
      "client_name": "Юрий Николов Величков",
      "egn": "6104266565",
      "floor": "1",
      "id": 9213,
      "mobile": "0889827210",
      "municipality_name": "Русе",
      "phone": "9294130",
      "postcode": "7000",
      "region_name": "РУСЕ",
      "representative": "",
      "street_name": "СЛАВЯНСКИ",
      "street_no": "14",
      "vhod": "1",
    },
    Object {
      "apartment": "2",
      "blok": "18",
      "car_ids": null,
      "city": "СИЛИСТРА",
      "client_name": "First Middle Last",
      "egn": "6310305565",
      "floor": "1",
      "id": 9219,
      "mobile": "0891043493",
      "municipality_name": "Силистра",
      "phone": "02830123",
      "postcode": "7500",
      "region_name": "СИЛИСТРА",
      "representative": "",
      "street_name": "ГЕНЕРАЛ  СКОБЕЛЕВ",
      "street_no": "",
      "vhod": "А",
    },
  ],
  "total": 1030889,
}
`;

exports[`Test persons data retrieval should get persons from db with sorting 1`] = `
Object {
  "data": Array [
    Object {
      "apartment": "0",
      "blok": "0",
      "car_ids": null,
      "city": "РУСЕ",
      "client_name": "Юрий Николов Величков",
      "egn": "6104266565",
      "floor": "1",
      "id": 9213,
      "mobile": "0889827210",
      "municipality_name": "Русе",
      "phone": "9294130",
      "postcode": "7000",
      "region_name": "РУСЕ",
      "representative": "",
      "street_name": "СЛАВЯНСКИ",
      "street_no": "14",
      "vhod": "1",
    },
    Object {
      "apartment": "2",
      "blok": "18",
      "car_ids": null,
      "city": "СИЛИСТРА",
      "client_name": "First Middle Last",
      "egn": "6310305565",
      "floor": "1",
      "id": 9219,
      "mobile": "0891043493",
      "municipality_name": "Силистра",
      "phone": "02830123",
      "postcode": "7500",
      "region_name": "СИЛИСТРА",
      "representative": "",
      "street_name": "ГЕНЕРАЛ  СКОБЕЛЕВ",
      "street_no": "",
      "vhod": "А",
    },
  ],
  "total": 1030889,
}
`;
